file/Upload.php

```diff
@@ -0,0 +1,234 @@
+<?php
+/**
+ * @link http://www.yiiframework.com/
+ * @copyright Copyright (c) 2008 Yii Software LLC
+ * @license http://www.yiiframework.com/license/
+ */
+
+namespace yii\mongodb\file;
+
+use MongoDB\BSON\Binary;
+use MongoDB\BSON\ObjectID;
+use MongoDB\BSON\UTCDatetime;
+use yii\base\InvalidParamException;
+use yii\base\Object;
+
+/**
+ * Upload represents the GridFS upload operation.
+ *
+ * @author Paul Klimov <klimov.paul@gmail.com>
+ * @since 2.1
+ */
+class Upload extends Object
+{
+    /**
+     * @var Collection file collection to be used.
+     */
+    public $collection;
+    /**
+     * @var string filename to be used for file storage.
+     */
+    public $filename;
+    /**
+     * @var array additional file document contents.
+     * Common GridFS columns:
+     *
+     * - metadata: array, additional data associated with the file.
+     * - aliases: array, an array of aliases.
+     * - contentType: string, content type to be stored with the file.
+     */
+    public $document = [];
+    /**
+     * @var integer chunk size in bytes.
+     */
+    public $chunkSize = 261120;
+    /**
+     * @var integer total upload length in bytes.
+     */
+    public $length = 0;
+    /**
+     * @var integer file chunk counts.
+     */
+    public $chunkCount = 0;
+
+    /**
+     * @var ObjectID file document ID.
+     */
+    protected $documentId;
+
+    /**
+     * @var resource has context for collecting md5 hash
+     */
+    private $hashContext;
+    /**
+     * @var string internal data buffer
+     */
+    private $buffer;
+
+
+    /**
+     * @inheritdoc
+     */
+    public function init()
+    {
+        $this->hashContext = hash_init('md5');
+
+        if (isset($this->document['_id'])) {
+            $this->documentId = $this->document['_id'] instanceof ObjectID ? $this->document['_id'] : ObjectID($this->document['_id']);
+        } else {
+            $this->documentId = new ObjectID();
+        }
+
+        $this->collection->ensureIndexes();
+    }
+
+    /**
+     * Adds string content to the upload.
+     * This method can invoked several times before [[complete()]] is called.
+     * @param string $content binary content.
+     * @return $this self reference.
+     */
+    public function addContent($content)
+    {
+        $freeBufferLength = $this->chunkSize - strlen($this->buffer);
```

Comment:
Should be `mb_strlen($this->buffer, '8bit')` in order to work properly if mb function overload is active.